lxd/device/gpu_physical.go

```diff
@@ -98,10 +101,209 @@ func (d *gpuPhysical) Start() (*deviceConfig.RunConfig, error) {
+			}
+
+			srcFDHandlers = append(srcFDHandlers, f)
+		} else {
+			err := os.Mkdir(devPath, 0700)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Mount the fs.
+		err = DiskMount(srcPath, devPath, false, "", mntOptions, fsName)
+		if err != nil {
+			return err
+		}
+
+		if isFile {
+			options = append(options, "create=file")
+		} else {
+			options = append(options, "create=dir")
+		}
+
+		runConf.Mounts = append(runConf.Mounts, deviceConfig.MountEntryItem{
+			DevName:    deviceName,
+			DevPath:    devPath,
+			TargetPath: relativeDestPath,
+			FSType:     "none",
+			Opts:       options,
+			OwnerShift: ownerShift,
+		})
+
+		runConf.PostHooks = append(runConf.PostHooks, func() error {
+			err := unix.Unmount(devPath, unix.MNT_DETACH)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		})
+	}
+
+	return nil
+}
+
+func (d *gpuPhysical) generateCDIHooksFilePath() string {
+	return filepath.Join(d.inst.DevicesPath(), fmt.Sprintf("%s%s", d.name, cdi.CDIHooksFileSuffix))
+}
+
 // startContainer detects the requested GPU devices and sets up unix-char devices.
 // Returns RunConfig populated with mount info required to pass the unix-char devices into the container.
 func (d *gpuPhysical) startContainer() (*deviceConfig.RunConfig, error) {
 	runConf := deviceConfig.RunConfig{}
+	if d.config["id"] != "" {
+		// Check if the id of the device match a CDI format.
+		cdiID, err := cdi.ToCDI(d.config["id"])
+		if err != nil {
+			return nil, err
+		}
+
+		// The cdiID can be empty if the id doesn't match the general CDI format, but this is not an error.
+		// and we allow the program to continue as the id could be a valid DRM card ID.
+		if !cdiID.Empty() {
+			if cdiID.Class == cdi.MIG {
+				return nil, fmt.Errorf(`MIG GPU notation detected for a "physical" gputype device. Choose a "mig" gputype device instead.`)
+			}
+
+			configDevices, hooks, err := cdi.GenerateFromCDI(d.inst, cdiID, d.logger)
+			if err != nil {
+				return nil, err
+			}
+
+			// Start the devices needed by the CDI specification.
+			err = d.startCDIDevices(cdiID, *configDevices, &runConf)
+			if err != nil {
+				return nil, err
+			}
+
+			// Persist the hooks to be run on a `lxc.hook.mount` LXC hook.
+			hooksFile := d.generateCDIHooksFilePath()
+			f, err := os.Create(hooksFile)
+			if err != nil {
+				return nil, fmt.Errorf("Could not create the CDI hooks file: %w", err)
+			}
+
+			defer f.Close()
+			err = json.NewEncoder(f).Encode(hooks)
+			if err != nil {
+				return nil, fmt.Errorf("Could not write to the CDI hooks file: %w", err)
+			}
+
+			// This is used to clean up the CDI hooks file on the postStop hook.
+			err = d.volatileSet(map[string]string{"cdi": "true"})
+			if err != nil {
+				return nil, err
+			}
+
+			runConf.GPUDevice = append(runConf.GPUDevice,
+				[]deviceConfig.RunConfigItem{
+					{Key: cdi.CDIHookDefinitionKey, Value: filepath.Base(hooksFile)},
+				}...)
+
+			return &runConf, nil
+		}
+	}
+
 	gpus, err := resources.GetGPU()
```

Comment:
Need a comment here explaining that we are proceeding to use the non-CDI approach.

Reply:
fixed